File: Data Structures & Algorithms/asteroid-collision/submission-2.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();

        for (int asteroid : asteroids) {
            boolean destroyed = false;

            // Collision only when: top is positive, current is negative
            while (!stack.isEmpty() && stack.peek() > 0 && asteroid < 0) {
                int top = stack.peek();
                int cur = Math.abs(asteroid);

                if (top < cur) {
                    // Stack top explodes, current survives → keep looping
                    stack.pop();
                } else if (top == cur) {
                    // Both explode
                    stack.pop();
                    destroyed = true;
                    break;
                } else {
                    // Current explodes, stack top survives
                    destroyed = true;
                    break;
                }
            }

            if (!destroyed) {
                stack.push(asteroid);
            }
        }

        // Stack is in reverse order — convert to array
        int[] result = new int[stack.size()];
        for (int i = result.length - 1; i >= 0; i--) {
            result[i] = stack.pop();
        }
        return result;
    }
}

File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-2.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if len(token) > 1 or token[0].isdigit():
                stack.append(int(token))
                continue
            x = stack.pop()
            if token == '+':
                stack[-1] += x
            elif token == '-':
                stack[-1] -= x
            elif token == '*':
                stack[-1] *= x 
            else:
                y = stack[-1]
                stack[-1] = y // x if y * x >= 0 else -(-y // x)
        return stack[0]